ARCMWDT: fix assorted build failures

tagunil opened the pull request:
At the moment we have a lot of failed tests on ARC platforms while using ARC MWDT toolchain. This PR is an attempt to fix at least the most obvious build failures before diving into runtime ones.

tagunil committed:
lib: libc: arcmwdt: replace the fileno macro

The fileno macro defined in ARC MWDT headers conflicts with the fileno()
function defined in lib/posix/options/device_io.c. We should undefine it
and replace with a POSIX-compliant declaration and a weak definition.

Signed-off-by: Ilya Tagunov <Ilya.Tagunov@synopsys.com>

tagunil committed:
lib: libc: arcmwdt: add declarations for reentrant time functions

ARC MWDT library does not contain implementations for the reentrant
POSIX time functions, so common Zephyr implementations end up being
used instead. However, there are no declarations, so we just add them
here, unguarded, until a better place is found.

Signed-off-by: Ilya Tagunov <Ilya.Tagunov@synopsys.com>

tagunil committed:
tests: lib: c_lib: add the proper dependency for asctime_r and ctime_r

As the test uses reentrant time functions like asctime_r or ctime_r,
it should select POSIX_C_LANG_SUPPORT_R. Otherwise there are build
failures observed with the default test configuration if these
functions are not provided by the toolchain.

Signed-off-by: Ilya Tagunov <Ilya.Tagunov@synopsys.com>

tagunil committed:
lib: libc: arcmwdt: move ssize_t definition from the toolchain header

Until POSIX layer decoupling is done, we still need the size_t type
definition across the codebase, but it should come through sys/types.h
and not through toolchain-specific headers.

Signed-off-by: Ilya Tagunov <Ilya.Tagunov@synopsys.com>

tagunil committed:
include: posix: remove an unreachable clockid_t definition

The clockid_t definition in question is not reachable anymore
because at least one of _CLOCKID_T_DECLARED and __clockid_t_defined
macros is already defined by that point, always making the check fail.

Signed-off-by: Ilya Tagunov <Ilya.Tagunov@synopsys.com>

tagunil committed:
lib: libc: arcmwdt: avoid double definition of clock_t and clockid_t

The clock_t and clockid_t types are already defined in ARC MWDT headers,
so they should not be defined again in Zephyr headers.

Signed-off-by: Ilya Tagunov <Ilya.Tagunov@synopsys.com>

keith-packard commented on include/zephyr/posix/posix_types.h:
```diff
@@ -84,11 +84,6 @@ typedef unsigned long useconds_t;
 #endif
 
 /* time related attributes */
-#if !defined(CONFIG_NEWLIB_LIBC) && !defined(CONFIG_ARCMWDT_LIBC)
```
This change doesn't seem to be relevant to making code work with ARCMWDT? It seems to be explicitly skipping this block when using that toolchain, after all?

tagunil replied:
Yeah, so that's a separate commit. It removes a block of code that became unreachable and redundant (see lines 20 to 24 of the same file). Still, if you believe this change is wrong, I'll drop the commit.

keith-packard replied:
Thanks for the explanation. Looks good to me.

cfriedt reviewed (dismissed):
Looks good to me - I guess it should probably be tagged for a back port once 4.0.0 is cut

tagunil committed:
kconfig: assume the host libc provides POSIX reentrant functions

This change is needed to avoid including Zephyr-specific implementations
of POSIX reentrant functions while building natively. One can safely
assume that the host libc indeed provides its own implementations of
the said functions.

Signed-off-by: Ilya Tagunov <Ilya.Tagunov@synopsys.com>

tagunil commented:
I've rebased the changes and added one more commit in the middle to prevent native_sim test failures. Specifically, I've made NATIVE_LIBC select TC_PROVIDES_POSIX_C_LANG_SUPPORT_R, as it looks like a reasonable assumption.

cfriedt commented on lib/libc/arcmwdt/libc-hooks.c:
```diff
@@ -71,6 +71,13 @@ int _write(int fd, const char *buf, unsigned int nbytes)
 }
 #endif
 
+#ifndef CONFIG_POSIX_DEVICE_IO
+__weak int fileno(FILE *file)
+{
+	return _fileno(file);
+}
+#endif
```
Does something like this work? I guess not many people have access to the arc mwdt toolchain to test 

```suggestion
__weak int _fileno(FILE *file)
{
	return zvfs_fileno(file);
}
```

tagunil replied:
Well, we have a `_fileno()` function in our ARC toolchain, but not a `fileno()` function (it's a macro). So my idea was:
- Undefine the macro in any case because it prevents doing anything
- If CONFIG_POSIX_DEVICE_IO is defined, just use `fileno()` from the POSIX subsystem
- Otherwise define `fileno()` as a wrapper around our `_fileno()` to avoid user breakage
I can't see how your proposal solves the issue of missing `fileno()`. But if you have any suggestions regarding the situation I've described above, they will be very welcome.

cfriedt replied:
`fileno()` shouldn't be missing if `CONFIG_POSIX_DEVICE_IO` is under.

tagunil replied:
Okay, so there are two separate questions: how to fix the issue of missing `fileno()` in the MWDT toolchain and whether we want to tie `_fileno()` to `zvfs_fileno()` unconditionally.
As to the first one, I still think what I did makes sense and I see no other obvious way to fix the issue.
But as to the second one, I'd like to hear your suggestions. Should we do this unconditionally or it should depend on something enabled? CONFIG_ZVFS? CONFIG_FDTABLE? Anything else?

tagunil replied:
And, to be honest, I don't understand at all how the `fileno()` and `fdopen()` functions provided by ZVFS could ever be plugged into e.g. `fwrite()` from any embedded libc, either MWDT or Picolibc, because of different interpretations of what a `FILE` is. It might require changing libc integration layers completely.

cfriedt replied:
> But as to the second one, I'd like to hear your suggestions. Should we do this unconditionally or it should depend on something enabled? CONFIG_ZVFS? CONFIG_FDTABLE? Anything else?


Good catch, an `ifdef CONFIG_FDTABLE` would probably work

tagunil replied:
Well, I've checked the source and found that the MWDT library doesn't use `_fileno()` internally, so there is no way to replace it with the ZVFS one. I'd really like to see someday a nice clean stdio sitting on top of the POSIX layer, but this is not the day yet. As for now, I propose to keep the change as is.

keith-packard replied:
> Does something like this work? I guess not many people have access to the arc mwdt toolchain to test

This is definitely wrong -- the FILE here is from libc, not the (invalid) FILE from zvfs.

cfriedt replied:
> I'd really like to see someday a nice clean stdio sitting on top of the POSIX layer, but this is not the day yet. 

That would be possible if Zephyr decided to become a POSIX OS.

Feel free to request that as an enhancement. I would support it.

Technically, POSIX is already mandatory for some types and Zephyr supports much of what is required already.

cfriedt replied:
> This is definitely wrong -- the FILE here is from libc, not the (invalid) FILE from zvfs.

If the libc `FILE*` contained a `void*` representing the Zephyr `fd_entry*`, then that would work.

tagunil commented:
Hi @keith-packard @cfriedt @stephanosio, could you please revisit this PR? The changes from the last review were quite small.

@cfriedt, I agree that the current approach to `fileno()` is not perfect, but the proper integration between libc and the POSIX layer is a much more complex topic than this one. For now I'll be happy to have it buildable.

keith-packard reviewed (approved):
I don't see any major issues with this PR.

evgeniy-paltsev commented on Kconfig.zephyr:
```diff
@@ -448,6 +448,7 @@ config CODING_GUIDELINE_CHECK
 config NATIVE_LIBC
 	bool
 	select FULL_LIBC_SUPPORTED
+	select TC_PROVIDES_POSIX_C_LANG_SUPPORT_R
```
So previously we didn't provide POSIX reentrant functions in case of native build? Or we provided implementation from Zephyr?

cfriedt replied:
Yes ? Hard to say - I don't often test on native_sim

tagunil replied:
> So previously we didn't provide POSIX reentrant functions in case of native build? Or we provided implementation from Zephyr?

It's complicated. The native implementation always provided the functions but our configuration system was not aware of the fact. It kinda worked, because in-tree users don't bother with selecting the right symbol anyway. But as soon as the libc test got its dependencies fixed, the issue was uncovered.
